### DISCUSSION TITLE
What is the use case for eTags?

### DISCUSSION THREAD

**Discussion by mblarsen** [imgproxy/imgproxy#1328] [Q&A]
1. An etag is generated from the result and included in the header
2. The client will send the etag to CDN
3. The CDN will pass on the etag to imgproxy
4. imgproxy will download source and process image
5. Generate a new etag
6. Compare etags
7. Return
8. CDN doesn't have to download image
9. Client doesn't have to download image

The save is on the client, making sure their image is cached while still being unchanged, but the imgproxy would still have to process every time even with CDN, right?

**Comment by DarthSim** [comment]
Hey @mblarsen!

ETags are generated by imgproxy from the processing options and the source image (or the ETag provided by the image source).

When imgproxy receives a request containing the `If-None-Match` header, it checks if processing options or the source image were changed. If nothing is changed, imgproxy skips the processing and responds with 304 (Not Modified).

The way clients and CDNs treat ETags depends on the client/CDN. But the most common behavior is the following:

On the client:

1. The client receives a response with the `ETag` header from imgproxy (or, actually, a CDN) and puts it in the cache.
2. If the client needs the same image again, it check the cache first. If the image's cache is not expired, it uses the image from the cache.
3. If the cached version is expired but not pruned yet, the client sends a request to the imgproxy (or CDN) setting the `If-None-Match` tag containing the ETag.
4. If the client receives a 304 (Not Modified) response, it marks the cached image as still valid and uses it.

On the CDN:

1. When the CDN receives an ETag from the origin (in our case – imgproxy), it saves it alongside the cached image.
2. When the CDN receives a request from a client, it first checks its cache. If the cached image is valid and ETags match, it responds with 304 (Not Modified).
3. If the cached image is valid but ETags don't match, the CDN responds with the cached image and the actual ETag.
4. If the cached image is expired but not pruned yet, the CDN makes a request to the origin setting the `If-None-Match` tag. If it receives a 304 (Not Modified) response, it marks the cached image as valid and repeats steps 2 and 3.
